### ISSUE TITLE
[1.10.2] Substitution alias causes ItemStack with null Item to be displayed in Creative menu

### ISSUE THREAD

**Issue by Choonster** [MinecraftForge/MinecraftForge#3127]
Forge version: 1.10.2-12.18.1.2035
Test case: [link](https://github.com/Choonster/ForgeSubstitutionAliasTestMod/blob/8bf308e70afbbeb89c63cc50d827a8ba6146c8d2/src/main/java/choonster/forgesubstitutionaliastestmod/ForgeSubstitutionAliasTestMod.java)
FML log: [link](https://gist.github.com/Choonster/2bb3da3893695201e9ce130f52f39e02)

Stack trace of `ItemStack` creation:
```
at net.minecraft.item.ItemStack.setItem(ItemStack.java:1072)
at net.minecraft.item.ItemStack.<init>(ItemStack.java:100)
at net.minecraft.item.ItemStack.<init>(ItemStack.java:96)
at net.minecraft.item.ItemStack.<init>(ItemStack.java:83)
at net.minecraft.block.BlockDirt.getSubBlocks(BlockDirt.java:63)
at net.minecraft.item.ItemBlock.getSubItems(ItemBlock.java:166)
at net.minecraft.creativetab.CreativeTabs.displayAllRelevantItems(CreativeTabs.java:309)
at net.minecraft.client.gui.inventory.GuiContainerCreative.setCurrentCreativeTab(GuiContainerCreative.java:487)
at net.minecraft.client.gui.inventory.GuiContainerCreative.initGui(GuiContainerCreative.java:276)
at net.minecraft.client.gui.GuiScreen.setWorldAndResolution(GuiScreen.java:549)
at net.minecraft.client.Minecraft.displayGuiScreen(Minecraft.java:1018)
at net.minecraft.client.gui.inventory.GuiInventory.initGui(GuiInventory.java:54)
at net.minecraft.client.gui.GuiScreen.setWorldAndResolution(GuiScreen.java:549)
at net.minecraft.client.Minecraft.displayGuiScreen(Minecraft.java:1018)
at net.minecraft.client.Minecraft.processKeyBinds(Minecraft.java:2207)
at net.minecraft.client.Minecraft.runTickKeyboard(Minecraft.java:2058)
at net.minecraft.client.Minecraft.runTick(Minecraft.java:1846)
at net.minecraft.client.Minecraft.runGameLoop(Minecraft.java:1118)
at net.minecraft.client.Minecraft.run(Minecraft.java:406)
at net.minecraft.client.main.Main.main(Main.java:118)
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.lang.reflect.Method.invoke(Method.java:498)
at net.minecraft.launchwrapper.Launch.launch(Launch.java:135)
at net.minecraft.launchwrapper.Launch.main(Launch.java:28)
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.lang.reflect.Method.invoke(Method.java:498)
at net.minecraftforge.gradle.GradleStartCommon.launch(GradleStartCommon.java:97)
at GradleStart.main(GradleStart.java:26)
```

The dirt `ItemBlock` is calling `Block#getSubBlocks` on the original `BlockDirt` instance, which creates several `ItemStack`s of `this`. `Item.getItemFromBlock` returns `null` for this instance, causing the `ItemStack`s to be created with a `null` `Item`.

**Comment by cpw** [comment]
Heh, the item mapping isn't updated..

**Comment by cpw** [comment]
Hmmm, actually, to fix this, I have to remove ItemBlock.block completely. It should be a reference resolved by the ITEM_BLOCK mapping. For now, you're going to have to work around it by substituting the ItemBlock too.

**Comment by Choonster** [comment]
Would that happen in 1.10.2, or is a breaking change that needs to wait for the next Minecraft version?

I don't really mind myself, the only place I'm using the substitution alias system is in this test mod.

**Comment by cpw** [comment]
Well, I'm trying to get this to work properly...

**Comment by diesieben07** [comment]
I'm not trying to be negative here, but I don't see how this is ever going to work properly. 
What if something that is not vanilla or Forge (i.e. something you can't change) has such a hard link like ItemBlock.block? There is no way to make that work.

**Comment by williewillus** [comment]
Concur, I think just requiring substitutions to the items as well is the best choice.

Blocks and itemblocks are actually pretty weakly linked, so this drawback is not a big one, Imo.

**Comment by diesieben07** [comment]
I was talking about substitutions in general :D

**Comment by LexManos** [comment]
Thats broken modder code then.
We have the mechanics in place {@ObjectHolder} to silently replace modder references, they just need to use it.
And as with all other Forge features if a modder does not use it, it's not our fault.

**Comment by diesieben07** [comment]
`@ObjectHolder` works for static references per registry name, yes. But not for dynamic associations (akin to `ItemBlock.block`.

**Comment by LexManos** [comment]
And thats what the remapping event is for.
You could potentially wrap @ObjectHolder to instance variables.
There are plenty of ways to do it and if you don't there isnt much we can do.

**Comment by Choonster** [comment]
This seems to be fixed by 4dcee246d50442123852c66b2351de1e9d533917. Should this issue be closed?

**Comment by TehNut** [comment]
On the topic of that commit, could an overload that defaults the last parameter to null be added so mods currently using `PersistentRegistryManager.createRegistry(...)` don't crash on latest Forge?
